### PR TITLE
[quantization] FP16 Dynamic Linear support

### PR DESCRIPTION
**Pull request by jamesr66a**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30045 [quantization] FP16 Dynamic Linear support**

Depends on FBGEMM PR https://github.com/pytorch/FBGEMM/pull/201

Differential Revision: [D18580648](https://our.internmc.facebook.com/intern/diff/D18580648)

### CODE REVIEW

File: torch/nn/quantized/dynamic/modules/linear.py [hunk 1]
```diff
@@ -66,10 +72,13 @@ def from_float(cls, mod):
...
-        weight_observer(mod.weight)
-        wt_scale, wt_zp = weight_observer.calculate_qparams()
-        qweight = torch.quantize_per_tensor(mod.weight.float(), float(wt_scale), int(wt_zp), torch.qint8)
-        qlinear = Linear(mod.in_features, mod.out_features)
+        dtype = weight_observer.dtype
+        if dtype == torch.qint8:
+            weight_observer(mod.weight)
+            wt_scale, wt_zp = weight_observer.calculate_qparams()
+            qweight = torch.quantize_per_tensor(mod.weight.float(), float(wt_scale), int(wt_zp), torch.qint8)
+        else:
```
Comment (jerryzh168) [inline torch/nn/quantized/dynamic/modules/linear.py]: should we do check for `else if dtype == torch.float` here as well?

File: torch/nn/quantized/modules/linear.py [hunk 2]
```diff
@@ -48,10 +48,13 @@ def __init__(self, in_features, out_features, bias_=True):
         if bias_:
             bias = torch.zeros(out_features, dtype=torch.float)
 
+        if dtype == torch.qint8:
+            qweight = torch._empty_affine_quantized(
+                [out_features, in_features], scale=1, zero_point=0, dtype=torch.qint8)
+        else:
```
Comment (jerryzh168) [inline torch/nn/quantized/modules/linear.py]: also here, check for `torch.float`?

File: torch/nn/quantized/modules/linear.py [hunk 3]
```diff
@@ -36,7 +36,7 @@ class Linear(torch.nn.Module):
     """
     _FLOAT_MODULE = nn.Linear
 
-    def __init__(self, in_features, out_features, bias_=True):
+    def __init__(self, in_features, out_features, bias_=True, dtype=torch.qint8):
```
Comment (jerryzh168) [inline torch/nn/quantized/modules/linear.py]: this seems to be changing the static quantization module, do we have static quantization kernel support for fp16 data?
Also can you add tests for this change and change the commit message to reflect this?

Comment (jamesr66a) [reply torch/nn/quantized/modules/linear.py]: So long as the dynamic quantization module inherits from this module, this module needs to have the APIs necessary to support dynamic quantization, or we can make the dynamic module not inherit from this one.

As it is, `forward()` of the static module should fail if fp16 packed weights are given to it

Comment (jerryzh168) [reply torch/nn/quantized/modules/linear.py]: can we split the common part to another module then?

File: torch/nn/quantized/modules/linear.py [hunk 4]
```diff
@@ -113,11 +120,17 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
...
 
+        self.dtype = state_dict[prefix + 'dtype']
+        state_dict.pop(prefix + 'dtype')
+
         super(Linear, self)._load_from_state_dict(state_dict, prefix, local_metadata, False,
                                                   missing_keys, unexpected_keys, error_msgs)
 
     @torch.jit.export
     def __setstate__(self, state):
+        # This needs to go before self.set_weight_bias
```
Comment (jianyuh) [inline torch/nn/quantized/modules/linear.py]: Any reason?

File: aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp [hunk 5]
```diff
@@ -217,14 +217,79 @@ class QLinearDynamicInt8 final : public torch::OperatorKernel {
...
+    TORCH_CHECK(input.dim() >= 2);
+
+    const int64_t M = size_to_dim_(input.dim() - 1, input.sizes());
+    const int64_t N = packed_weight_fp16.numCols();
+    std::vector<int64_t> output_size = input.sizes().vec();
+    output_size.back() = N;
+    Tensor output = at::empty(output_size, input.options().dtype(at::kFloat));
+
+    // Call the fp16 gemm interface
+    fbgemm::cblas_gemm_compute(
```
Comment (jianyuh) [inline aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp]: Do you also want to add the intra-op parallelization support in this PR? Up to you. I can add a follow-up PR for that soon.

**Review by jianyuh — COMMENTED** [review]
Overall it looks great! Thanks for enabling this!

File: aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp (hunk 5 above)
Comment (jamesr66a) [reply aten/src/ATen/native/quantized/cpu/qlinear_dynamic.cpp]: I think that's better as a follow-up

File: torch/nn/quantized/dynamic/modules/linear.py [hunk 6]
```diff
@@ -27,17 +27,23 @@ class Linear(nnq.Linear):
...
         # We don't muck around with buffers or attributes or anything here
         # to keep the module simple. *everything* is simply a Python attribute.
         # Serialization logic is explicitly handled in the below serialization and
         # deserialization modules
 
     def forward(self, x):
         # Note that we can handle self.bias == None case.
-        Y = torch.ops.quantized.linear_dynamic(
-            x, self._packed_params)
+        if self.dtype == torch.qint8:
```
[the remark this answers is not in the document]
Comment (jamesr66a) [reply torch/nn/quantized/dynamic/modules/linear.py]: I think the existing convention is the qint types only, right?

File: torch/nn/quantized/modules/linear.py (hunk 4 above)
Comment (jamesr66a) [reply torch/nn/quantized/modules/linear.py]: Yes, otherwise `self.dtype` will be uninitialize and `set_weight_bias` cannot check that value

Comment (jianyuh) [reply torch/nn/quantized/modules/linear.py]: Might be good to add that in the comments.

**Review by jianyuh — APPROVED** [review]
LGTM! Thanks for completing this! Please make sure to address @jerryzh168 's comments.

**Comment by jamesr66a** [comment]
@jianyuh actually this patch is not going to cleanly merge with master, so I'm going to have to do some work to rebase it sometime

**Pull request closed without merging**